src/TranslateTypeDeclarations.re

```diff
@@ -36,14 +36,17 @@ let createCase = ((label, attributes)) =>
 
 // Rename record fields.
 // If @genType.as is used, perform renaming conversion.
-// If @bs.as is used (with records-as-objects active), no conversion is required.
+// If @bs.as is used (with records-as-objects active), quote and escape
+// because it's likely that @bs.as was used to allow invalid identifier characters
 let renameRecordField = (~config, ~attributes, ~nameRE) => {
   switch (attributes |> Annotation.getGenTypeAsRenaming) {
   | Some(nameJS) => (nameJS, nameRE)
   | None =>
     if (config.recordsAsObjects) {
       switch (attributes |> Annotation.getBsAsRenaming) {
-      | Some(name) => (name, name)
+      | Some(nameBS) =>
+        let nameJS = nameBS |> String.escaped |> EmitText.quotes;
```

Comment:
How about a simple check if it needs escaping?
Turns out this exists already inside `escaped` in `src/ext/ext_string.ml` and can be brought out.

Reply:
Doesn't `String.escaped` already check if it needs escaping before actually applying the escaping?

Did you mean a check if it needs quoting? The heuristics around which identifiers need quoting in JS are incredibly complicated and I was going down a rabbit hole trying to build a reliable algorithm. The easiest seemed just to quote anything that had been renamed with `@bs.as`, because it's likely this was done for the same reason (i.e. the identifier is invalid in Reason too). The only situation I can think of where one would use `@bs.as` for an identifier that _doesn't_ need quoting in JS is for Reason keywords such as `type` – maybe we should keep a list of keywords and check against that? But I thought it was ok to just quote everything, given that most people won't look at their type files for too long and intellisense (in VS Code) removes unnecessary quotes anyway when hovering over types.

Reply:
I would rather not quote everything. Two simple approximations are:
1)  if it needs escaping, add quotes. It's not perfect, but a good first approx we can live with, or
2) if it is all alphanumeric characters, it does not need quoting (I guess, maybe with a small refinement), any other case add quotes.